include/adaflo/sharp_interface_util.h

```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
+                                    dummy,
+                                    surface_quadrature,
+                                    update_quadrature_points);
 
-  std::vector<double>                  buffer;
-  std::vector<double>                  buffer_dim;
-  std::vector<types::global_dof_index> local_dof_indices;
+    for (const auto &cell : surface_mesh.active_cell_iterators())
+      {
+        if (cell->is_locally_owned() == false)
+          continue;
 
-  for (unsigned int i = 0; i < cells.size(); ++i)
-    {
-      typename DoFHandler<dim>::active_cell_iterator cell = {
-        &dof_handler.get_triangulation(), cells[i].first, cells[i].second, &dof_handler};
+        fe_eval.reinit(cell);
 
-      typename DoFHandler<dim>::active_cell_iterator cell_dim = {
-        &dof_handler.get_triangulation(),
-        cells[i].first,
-        cells[i].second,
-        &dof_handler_dim};
+        for (const auto q : fe_eval.quadrature_point_indices())
+          integration_points.push_back(fe_eval.quadrature_point(q));
+      }
 
-      local_dof_indices.resize(cell->get_fe().n_dofs_per_cell());
-      buffer.resize(cell->get_fe().n_dofs_per_cell());
-      buffer_dim.resize(cell->get_fe().n_dofs_per_cell() * dim);
+    return integration_points;
+  }();
 
-      cell->get_dof_indices(local_dof_indices);
+  Utilities::MPI::RemotePointEvaluation<spacedim, spacedim> eval;
+  eval.reinit(integration_points, dof_handler.get_triangulation(), mapping);
 
-      const unsigned int n_points = ptrs[i + 1] - ptrs[i];
+  const auto integration_values = [&]() {
+    std::vector<T> integration_values;
 
-      const ArrayView<const Point<dim>> unit_points(points.data() + ptrs[i], n_points);
-      const ArrayView<const double>     JxW(weights.data() + ptrs[i], n_points);
+    FE_Nothing<dim, spacedim> dummy;
 
-      // gather curvature
-      constraints.get_dof_values(curvature_solution,
-                                 local_dof_indices.begin(),
-                                 buffer.begin(),
-                                 buffer.end());
+    FEValues<dim, spacedim> fe_eval(surface_mapping,
+                                    dummy,
+                                    surface_quadrature,
+                                    update_JxW_values);
 
-      // evaluate curvature
-      phi_curvature.evaluate(cell,
-                             unit_points,
-                             make_array_view(buffer),
-                             EvaluationFlags::values);
+    for (const auto &cell : surface_mesh.active_cell_iterators())
+      {
+        if (cell->is_locally_owned() == false)
+          continue;
 
-      // gather normal
-      for (int i = 0; i < dim; ++i)
+        fe_eval.reinit(cell);
+
+        for (const auto q : fe_eval.quadrature_point_indices())
+          integration_values.push_back(fe_eval.JxW(q));
+      }
+
+    return integration_values;
+  }();
+
+  const auto fu = [&](const auto &values, const auto &cell_data) {
```

Comment:
maybe you can find a more verbose name than `fu` :-)

Reply:
Why does no one like the function name (see also: https://github.com/dealii/dealii/pull/11804#discussion_r595248782)? I guess you want to point out that I should instead use `foo`(and `bar`) ;) Instead, I went with `integration_function`.